Update SocketPlatform setup strategy

kealjones-wk opened the pull request:
# Issue
An issue was discovered today that prevents socketcluster_client from being used with js/web based projects due to dart 2 now rejecting `import 'dart:io';` being in a web project at all even when only being imported transitively and no code for it was executed, which was previously done in order to use the `IoSocketPlatform` type in `socket.dart`.

# Fix
- Isolate the `dart:io` and `dart:html` import containing files behind conditional imports.
- Add `SocketPlatform` and other Base Types into separate `interface` class so they can be used for type checking.
- Add initial assignment to `_globalSocketPlatform` through a conditional import `RuntimeSocketPlatform` constant in the respective file, this makes the platform be auto configured! 
  - Web users will no longer need to add `globalSocketPlatform = httpSocketPlatform;` as the conditional import does this for them.

# QA
- [ ] Flutter based projects continue to work as expected. (@pauldemarco any chance you could test this?)
- [x] Web projects no longer have build time issues due to import. (confirmed by @kealjones-wk  with an over_react project)

kealjones-wk committed:
update socket platform setup strategy

kealjones-wk committed:
add export of interface outside of conditional

kealjones-wk committed:
Update socket_platform.dart

kealjones-wk committed:
cleanup

kealjones-wk commented:
@pauldemarco This is ready for your review, dart decided to change things on us so the old method no longer works for web projects, i am unsure about vm/flutter projects, but this PR should remove the issue for everyone hopefully. If you have any questions/concerns, let me know :) Thanks!

kealjones-wk committed:
remove leftover print statment

kealjones-wk committed:
fix extra space before `;`

MichaelMarner commented:
I'm just testing this against [Remote Devtools](https://github.com/MichaelMarner/dart-redux-remote-devtools) and have found issues trying to get the Socket to connect on Mobile.

I believe Remote Devtools is currently broken on mobile platforms

MichaelMarner commented:
@kealjones-wk I am about to make a PR to your fork which fixes the platform selection between web and dart.io - looks like it has been broken for a while. If you merge my PR into your branch, we should be good

MichaelMarner committed:
Fix switching between dart.io and dart.html websockets

The changes made to socketcluster_client to support web platform broke mobile, as the checks were implemented incorrectly.

* Need to use `is` to test runtime types, `==` and `!=` do not work.
* `IoSocketPlatform.webSocket` was never actually calling connect

Closes #11

kealjones-wk committed:
Merge pull request #1 from MichaelMarner/auto-platform-selection

Fix switching between dart.io and dart.html websockets

kealjones-wk commented:
@MichaelMarner got your changes merged in and tested on web all is well on my end, sorry again if my previous pr unknowingly broke your repo [MichaelMarner/dart-redux-remote-devtools](https://github.com/MichaelMarner/dart-redux-remote-devtools), haha that was the opposite of what I wanted I was trying to make it usable for dart web projects with [Workiva/over_react](https://github.com/Workiva/over_react). We've got a react-redux wrapper in the works and your lib + this are essential to the dev flow :)

MichaelMarner commented:
This should be fine to merge now - I have verified it still works in Flutter mobile, and @kealjones-wk has verified things work in web.

pauldemarco commented:
LGTM

pauldemarco commented:
New version 0.2.0 has been published to pubdev. https://pub.dev/packages/socketcluster_client
